posthog/session_recordings/queries/sub_queries/events_subquery.py

```diff
@@ -0,0 +1,260 @@
+from datetime import datetime, timedelta
+from typing import Optional
+
+from posthog.constants import PropertyOperatorType
+from posthog.hogql import ast
+from posthog.hogql.property import property_to_expr
+from posthog.hogql.query import execute_hogql_query
+from posthog.hogql_queries.legacy_compatibility.filter_to_query import legacy_entity_to_node, MathAvailability
+from posthog.models import Team, Action, Entity
+from posthog.schema import (
+    RecordingsQuery,
+    HogQLQueryModifiers,
+    NodeKind,
+    PropertyGroupFilterValue,
+    FilterLogicalOperator,
+)
+from posthog.session_recordings.queries.session_recording_list_from_query import (
+    _entity_to_expr,
+    poe_is_active,
+    SessionRecordingQueryResult,
+    NEGATIVE_OPERATORS,
+    INVERSE_OPERATOR_FOR,
+    is_event_property,
+    is_group_property,
+    is_person_property,
+)
+from posthog.session_recordings.queries.sub_queries.base_query import SessionRecordingsListingBaseQuery
+
+
+class ReplayFiltersEventsSubQuery(SessionRecordingsListingBaseQuery):
+    def __init__(
+        self,
+        team: Team,
+        query: RecordingsQuery,
+        hogql_query_modifiers: Optional[HogQLQueryModifiers] = None,
+    ):
+        super().__init__(team, query)
+        self._hogql_query_modifiers = hogql_query_modifiers
+
+    @property
+    def _event_predicates(self):
+        event_exprs: list[ast.Expr] = []
+        event_names: set[int | str] = set()
+
+        for entity in self.entities:
+            if entity.kind == NodeKind.ACTIONS_NODE:
+                action = Action.objects.get(pk=int(entity.id), team__project_id=self._team.project_id)
+                event_names.update([ae for ae in action.get_step_events() if ae and ae not in event_names])
+            else:
+                if entity.event and entity.event not in event_names:
+                    event_names.add(entity.event)
+
+            entity_exprs = [_entity_to_expr(entity=entity)]
+
+            if entity.properties:
+                entity_exprs.append(property_to_expr(entity.properties, team=self._team, scope="replay_entity"))
+
+            event_exprs.append(ast.And(exprs=entity_exprs))
+
+        return event_exprs, list(event_names)
+
+    def _select_from_events(self, select_expr: ast.Expr) -> ast.SelectQuery:
+        return ast.SelectQuery(
+            select=[select_expr],
+            select_from=ast.JoinExpr(
+                table=ast.Field(chain=["events"]),
+            ),
+            where=self._where_predicates(),
+            having=self._having_predicates(),
+            group_by=[ast.Field(chain=["$session_id"])],
+            order_by=[ast.OrderExpr(expr=ast.Call(name="min", args=[ast.Field(chain="timestamp")]), order="DESC")],
```

Comment:
this subquery was extracted from the large file... and then this order by was added